### PR TITLE
Fix swap endpoints for looped flows

### LINKED ISSUE #3846: flowLoop technical debt
- [x] swapEndpoint for a flow with fowLoop on it
- [x] flowLoop for a QinQ flow
- [x] flowLoop in multiTable mode 
- [x] flowLoop for a single-switch flows
- [x] flowPing ? (it works because _'flowPing'_ doesn't verify terminating switch)
- [x] system shouldn't allow to create flowLoop twice on the src/dst for the same flow
- [ ] flowLoop for a vxlan flow
- [x] missed cache for flowLoop rules

### PR DESCRIPTION
**Pull request by rozdy**
Close #3846

### CODE REVIEW

**Review by dpoltavets — APPROVED** [review]

**Review by niksv — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/UpdateRequestAction.java [hunk 1]
```diff
@@ -54,6 +54,12 @@ private void sendUpdateCommand(Flow flow, RequestedFlow targetFlow, String anoth
         flowRequest.setDestination(
                 new FlowEndpoint(targetFlow.getDestSwitch(), targetFlow.getDestPort(), targetFlow.getDestVlan()));
 
+        if (flow.getLoopSwitchId() != null) {
+            boolean flowLoopedOnSrc = flow.getLoopSwitchId().equals(flow.getSrcSwitchId());
+            flowRequest.setLoopSwitchId(flowLoopedOnSrc ? flowRequest.getSource().getSwitchId()
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/swapendpoints/actions/UpdateRequestAction.java:59]: The idea behind this logic was to save flow in "src-looped" or "dest-looped" state regardless of what switches was source and destination. For flows with loop on transit switches swap operation is undefined for now.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/RequestedFlowMapper.java [hunk 2]
```diff
@@ -167,6 +167,7 @@ public PathComputationStrategy mapComputationStrategy(String raw) {
     @Mapping(target = "priority", source = "priority")
     @Mapping(target = "pinned", source = "pinned")
     @Mapping(target = "detectConnectedDevices", source = "detectConnectedDevices")
+    @Mapping(target = "loopSwitchId", source = "loopSwitchId")
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/RequestedFlowMapper.java:170]: Yes, I know but here all other fields was listed so for similarity reasons I decided to put it here.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowUpdateService.java [hunk 3]
```diff
@@ -143,8 +141,12 @@ public void handleCreateFlowLoopRequest(String key, CommandContext commandContex
         Optional<Flow> flow = flowRepository.findById(request.getFlowId());
         if (flow.isPresent()) {
             FlowRequest flowRequest = RequestedFlowMapper.INSTANCE.toFlowRequest(flow.get());
-            flowRequest.setLoopSwitchId(request.getSwitchId());
-            handleRequest(key, commandContext, flowRequest);
+            if (flowRequest.getLoopSwitchId() == null) {
+                flowRequest.setLoopSwitchId(request.getSwitchId());
+                handleRequest(key, commandContext, flowRequest);
+            } else {
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowUpdateService.java:147]: Fixed.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowUpdateService.java [hunk 4]
```diff
@@ -205,6 +207,13 @@ private Message buildFlowNotFoundErrorMessage(String flowId, CommandContext comm
         return new ErrorMessage(error, commandContext.getCreateTime(), commandContext.getCorrelationId());
     }
 
+    private Message buildFlowAlreadyLoopedErrorMessage(FlowRequest flow, CommandContext commandContext) {
+        String description = String.format("Flow is already looped on switch '%s'", flow.getLoopSwitchId());
+        ErrorData error = new ErrorData(ErrorType.UNPROCESSABLE_REQUEST,
+                String.format("Can't create flow loop on '%s' found", flow.getFlowId()), description);
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowUpdateService.java]: Fixed.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java [hunk 5]
```diff
@@ -105,22 +105,18 @@ public void validate(RequestedFlow flow, Set<String> bulkUpdateFlowIds)
     public void validate(Flow flow, RequestedFlow requestedFlow, Set<String> bulkUpdateFlowIds)
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java]: Fixed.

File: src-java/stats-topology/stats-storm-topology/src/main/java/org/openkilda/wfm/topology/stats/metrics/FlowMetricGenBolt.java [hunk 6]
```diff
@@ -64,7 +65,8 @@ protected void handleInput(Tuple input) throws Exception {
 
     private void emit(FlowStatsEntry entry, long timestamp, @Nonnull SwitchId switchId,
                       @Nullable CacheFlowEntry flowEntry) throws FlowCookieException {
-        boolean isFlowSegmentEntry = new Cookie(entry.getCookie()).getType() == CookieType.SERVICE_OR_FLOW_SEGMENT;
+        boolean isFlowSegmentEntry = new Cookie(entry.getCookie()).getType() == CookieType.SERVICE_OR_FLOW_SEGMENT
+                && !new FlowSegmentCookie(entry.getCookie()).isLooped();
```
Comment (nikitamarchenko) [inline src-java/stats-topology/stats-storm-topology/src/main/java/org/openkilda/wfm/topology/stats/metrics/FlowMetricGenBolt.java:69]: ugly look

**Review by nikitamarchenko — APPROVED** [review]

**Pull request merged**